mountpoint-s3-client/src/s3_crt_client/get_object.rs

```diff
@@ -19,189 +16,177 @@ use crate::object_client::{
+                .map_err(S3RequestError::construction_failure)?;
 
-        let request = self.inner.make_meta_request_from_options(
-            options,
-            span,
-            |_| (),
-            move |headers, status| {
-                // Headers can be returned multiple times, but the metadata/checksums don't change.
-                // Explicitly ignore the case where we've already set object metadata.
+            let mut options = S3CrtClientInner::new_meta_request_options(message, S3Operation::GetObject);
+            options.part_size(self.inner.read_part_size as u64);
+
+            let mut headers_sender = Some(headers_sender);
+
+            self.inner.make_meta_request_from_options(
+                options,
+                span,
+                |_| (),
+                move |headers, status| {
+                    // Headers can be returned multiple times, but the metadata/checksums don't change.
+                    // Explicitly ignore the case where we've already set object metadata.
+
+                    // Only set headers if we have a 2xx status code. If we only get other status codes,
+                    // then on_finish sets an error.
+                    if (200..300).contains(&status) {
+                        if let Some(sender) = headers_sender.take() {
+                            // Only send the headers once.
+                            _ = sender.send(headers.clone());
+                        }
+                    }
+                },
+                move |offset, data| {
+                    let _ = part_sender.unbounded_send((offset, data.into()));
+                },
+                move |result| {
+                    if result.is_err() {
+                        Err(parse_get_object_error(result).map(ObjectClientError::ServiceError))
+                    } else {
+                        Ok(())
+                    }
+                },
+            )?
+        };
 
-                // Only set headers if we have a 2xx status code. If we only get other status codes,
-                // then on_finish sets an error.
-                if (200..300).contains(&status) {
-                    // Don't overwrite if already set - the first headers are fine.
-                    object_headers_setter_on_headers.or_set(Ok(headers.clone()));
-                }
-            },
-            move |offset, data| {
-                let _ = sender.unbounded_send(Ok((offset, data.into())));
-            },
-            move |result| {
-                // FIXME - Ideally we'd include a reason why we failed here.
-                object_headers_setter_on_finish.or_set(Err(ObjectHeadersError::UnknownError));
-                if result.is_err() {
-                    Err(parse_get_object_error(result).map(ObjectClientError::ServiceError))
-                } else {
-                    Ok(())
-                }
-            },
-        )?;
+        let headers = select_biased! {
```

Comment:
It does not block because the CRT will initiate the request anyway. I've repurposed the `test_get_object_user_metadata_with_zero_backpressure` test to ensure that it works.